File: apps/api/src/modules/runtime-paths.ts
import * as fs from "node:fs";
import * as path from "node:path";

interface RuntimePathsRaw {
  databases?: {
    tournaments?: string[];
    mapStartDetection?: string;
  };
  artifacts?: {
    jobsStore?: string;
    tracksDir?: string;
    tracksFile?: string;
    mapAdminSettings?: string;
    zonesDir?: string;
    textZonesDir?: string;
  };
  media?: {
    recordsDir?: string;
    mapsDir?: string;
  };
}

export interface RuntimePaths {
  databases: {
    tournaments: string[];
    mapStartDetection: string;
  };
  artifacts: {
    jobsStore: string;
    tracksDir: string;
    tracksFile: string;
    mapAdminSettings: string;
    zonesDir: string;
    textZonesDir: string;
  };
  media: {
    recordsDir: string;
    mapsDir: string;
  };
}

const DEFAULTS: RuntimePaths = {
  databases: {
    tournaments: ["output/tournaments.sqlite", "output/youtube_ingest/tournaments.sqlite"],
    mapStartDetection: "output/map_start_detection.sqlite",
  },
  artifacts: {
    jobsStore: "output/jobs.json",
    tracksDir: "output/tracks",
    tracksFile: "output/tracks.json",
    mapAdminSettings: "output/admin_map_settings.json",
    zonesDir: "output/zones",
    textZonesDir: "output/text_zones",
  },
  media: {
    recordsDir: "ffmpeg_downloader/records",
    mapsDir: "maps",
  },
};

function resolvePath(projectRoot: string, relOrAbs: string): string {
  return path.isAbsolute(relOrAbs) ? relOrAbs : path.join(projectRoot, relOrAbs);
}

export function loadRuntimePaths(projectRoot: string): RuntimePaths {
  const cfgPath = path.join(projectRoot, "config", "runtime_paths.json");
  let raw: RuntimePathsRaw = {};
  if (fs.existsSync(cfgPath)) {
    try {
      raw = JSON.parse(fs.readFileSync(cfgPath, "utf-8")) as RuntimePathsRaw;
    } catch {
      raw = {};
    }
  }

  const tournamentsRel = raw.databases?.tournaments?.length ? raw.databases.tournaments : DEFAULTS.databases.tournaments;
  const runtime: RuntimePaths = {
    databases: {
      tournaments: tournamentsRel.map((item) => resolvePath(projectRoot, item)),
      mapStartDetection: resolvePath(projectRoot, raw.databases?.mapStartDetection ?? DEFAULTS.databases.mapStartDetection),
    },
    artifacts: {
      jobsStore: resolvePath(projectRoot, raw.artifacts?.jobsStore ?? DEFAULTS.artifacts.jobsStore),
      tracksDir: resolvePath(projectRoot, raw.artifacts?.tracksDir ?? DEFAULTS.artifacts.tracksDir),
      tracksFile: resolvePath(projectRoot, raw.artifacts?.tracksFile ?? DEFAULTS.artifacts.tracksFile),
      mapAdminSettings: resolvePath(projectRoot, raw.artifacts?.mapAdminSettings ?? DEFAULTS.artifacts.mapAdminSettings),
      zonesDir: resolvePath(projectRoot, raw.artifacts?.zonesDir ?? DEFAULTS.artifacts.zonesDir),
      textZonesDir: resolvePath(projectRoot, raw.artifacts?.textZonesDir ?? DEFAULTS.artifacts.textZonesDir),
    },
    media: {
      recordsDir: resolvePath(projectRoot, raw.media?.recordsDir ?? DEFAULTS.media.recordsDir),
      mapsDir: resolvePath(projectRoot, raw.media?.mapsDir ?? DEFAULTS.media.mapsDir),
    },
  };
  return runtime;
}

File: services/analysis/app/runtime_paths.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any


def _resolve(project_root: Path, raw: str) -> Path:
    value = Path(raw)
    return value if value.is_absolute() else (project_root / value)


def load_runtime_paths(project_root: Path) -> dict[str, Any]:
    cfg_path = project_root / "config" / "runtime_paths.json"
    payload: dict[str, Any] = {}
    if cfg_path.exists():
        try:
            payload = json.loads(cfg_path.read_text(encoding="utf-8"))
        except Exception:
            payload = {}

    db_cfg = payload.get("databases", {}) if isinstance(payload, dict) else {}
    art_cfg = payload.get("artifacts", {}) if isinstance(payload, dict) else {}
    media_cfg = payload.get("media", {}) if isinstance(payload, dict) else {}

    tournaments_raw = db_cfg.get("tournaments")
    if not isinstance(tournaments_raw, list) or not tournaments_raw:
        tournaments_raw = ["output/tournaments.sqlite", "output/youtube_ingest/tournaments.sqlite"]

    resolved_tournaments = [_resolve(project_root, str(item)) for item in tournaments_raw]
    preferred_tournaments = next((p for p in resolved_tournaments if p.exists()), resolved_tournaments[0])

    return {
        "databases": {
            "tournaments": resolved_tournaments,
            "preferred_tournaments": preferred_tournaments,
            "map_start_detection": _resolve(project_root, str(db_cfg.get("mapStartDetection", "output/map_start_detection.sqlite"))),
        },
        "artifacts": {
            "jobs_store": _resolve(project_root, str(art_cfg.get("jobsStore", "output/jobs.json"))),
            "tracks_dir": _resolve(project_root, str(art_cfg.get("tracksDir", "output/tracks"))),
            "tracks_file": _resolve(project_root, str(art_cfg.get("tracksFile", "output/tracks.json"))),
            "map_admin_settings": _resolve(project_root, str(art_cfg.get("mapAdminSettings", "output/admin_map_settings.json"))),
            "zones_dir": _resolve(project_root, str(art_cfg.get("zonesDir", "output/zones"))),
            "text_zones_dir": _resolve(project_root, str(art_cfg.get("textZonesDir", "output/text_zones"))),
        },
        "media": {
            "records_dir": _resolve(project_root, str(media_cfg.get("recordsDir", "ffmpeg_downloader/records"))),
            "maps_dir": _resolve(project_root, str(media_cfg.get("mapsDir", "maps"))),
        },
    }
